File: test/technology/sidb/simulation/analysis/test_minimum_energy.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/technology/sidb/simulation/analysis/minimum_energy.hpp`.
 * @author Jan Drewniok (Drewniok)
 * @author Marcel Walter (marcelwa)
 * @author Willem Lambooy (wlambooy)
 */

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>

#include <fiction/technology/sidb/charge_distribution.hpp>
#include <fiction/technology/sidb/layout.hpp>
#include <fiction/technology/sidb/model/charge_state.hpp>
#include <fiction/technology/sidb/simulation/analysis/minimum_energy.hpp>
#include <fiction/technology/sidb/simulation/potential_landscape.hpp>
#include <fiction/technology/sidb/technology.hpp>

#include <cmath>
#include <vector>

using namespace fiction;
using namespace fiction::sidb;
using namespace fiction::sidb::model;
using namespace fiction::sidb::simulation;
using namespace fiction::sidb::simulation::analysis;

TEST_CASE("Test minimum energy function", "[minimum-energy]")
{
    layout lyt{};

    SECTION("layout with no SiDB placed")
    {
        const potential_landscape        land{lyt};
        std::vector<charge_distribution> all_cds{};

        CHECK(std::isinf(minimum_energy(all_cds.begin(), all_cds.end())));

        all_cds.push_back(land.evaluate(charge_distribution{lyt}));

        CHECK(std::abs(minimum_energy(all_cds.begin(), all_cds.end()) - 0) < 0.00000001);
    }

    SECTION("layout with one SiDB placed")
    {
        lyt.assign_cell_type({0, 0, 0}, sidb_technology::cell_type::NORMAL);

        const potential_landscape        land{lyt};
        std::vector<charge_distribution> all_cds{};

        CHECK(std::isinf(minimum_energy(all_cds.cbegin(), all_cds.cend())));

        all_cds.push_back(land.evaluate(charge_distribution{lyt}));

        CHECK(std::abs(minimum_energy(all_cds.cbegin(), all_cds.cend()) - 0) < 0.00000001);
    }

    SECTION("layout with three SiDBs placed")
    {
        lyt.assign_cell_type({0, 0, 0}, sidb_technology::cell_type::NORMAL);
        lyt.assign_cell_type({10, 5, 0}, sidb_technology::cell_type::NORMAL);
        lyt.assign_cell_type({9, 4, 1}, sidb_technology::cell_type::NORMAL);

        const potential_landscape        land{lyt};
        std::vector<charge_distribution> all_cds{};

        CHECK(std::isinf(minimum_energy(all_cds.cbegin(), all_cds.cend())));

        charge_distribution first{lyt};
        first.assign_charge_state({0, 0, 0}, charge_state::NEUTRAL);
        all_cds.push_back(land.evaluate(first));

        charge_distribution second{lyt};
        second.assign_charge_state({10, 5, 0}, charge_state::NEUTRAL);
        second.assign_charge_state({9, 4, 1}, charge_state::NEUTRAL);
        all_cds.push_back(land.evaluate(second));

        CHECK_THAT(minimum_energy(all_cds.cbegin(), all_cds.cend()), Catch::Matchers::WithinAbs(0.0, 0.00001));
        CHECK(minimum_energy_distribution(all_cds.cbegin(), all_cds.cend())->same_charge_states(second));
    }
}

File: bindings/mnt/pyfiction/test/technology/sidb/simulation/analysis/test_minimum_energy.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Tests for ``minimum_energy``."""

from __future__ import annotations

import pytest

from mnt.pyfiction import (
    charge_distribution,
    lattice,
    lattice_site,
    minimum_energy,
    potential_landscape,
    sidb_charge_state,
    sidb_layout,
    sidb_simulation_parameters,
    sidb_technology,
)


@pytest.mark.parametrize(
    "lat",
    [pytest.param(lattice.si_100_2x1(), id="100"), pytest.param(lattice.si_111_1x1(), id="111")],
)
def test_three_sidbs(lat):
    layout = sidb_layout(lat)

    layout.assign_cell_type(lattice_site(0, 0, 1), sidb_technology.cell_type.NORMAL)
    layout.assign_cell_type(lattice_site(4, 0, 1), sidb_technology.cell_type.NORMAL)
    layout.assign_cell_type(lattice_site(6, 0, 1), sidb_technology.cell_type.NORMAL)

    landscape = potential_landscape(layout, sidb_simulation_parameters())

    cd1 = landscape.evaluate(charge_distribution(layout))  # all negative
    cd2 = landscape.evaluate(charge_distribution(layout, sidb_charge_state.NEUTRAL))  # all neutral
    cd3 = charge_distribution(layout)
    cd3.assign_charge_state(lattice_site(6, 0, 1), sidb_charge_state.NEUTRAL)  # only two SiDBs are negative
    cd3 = landscape.evaluate(cd3)

    cd4 = charge_distribution(layout)
    cd4.assign_charge_state(lattice_site(0, 0, 1), sidb_charge_state.NEUTRAL)
    cd4.assign_charge_state(lattice_site(4, 0, 1), sidb_charge_state.POSITIVE)
    cd4.assign_charge_state(lattice_site(6, 0, 1), sidb_charge_state.NEGATIVE)
    cd4 = landscape.evaluate(cd4)

    result = minimum_energy([cd1, cd2, cd3, cd4])

    assert result == pytest.approx(min(cd.energy() for cd in [cd1, cd2, cd3, cd4]))
    assert result == pytest.approx(cd4.energy())
    assert result < 0.0
